### PR TITLE
Prevent IDE crash in case of file duplicate in fsproj

### PR DESCRIPTION
**Pull request by T-Gro**
If the same file was listed twice for compilation (typically = text-based merge conflict resolution), the following happened:
- Build of that project worked fine
- IDE for that project worked fine
- Build for referencing projects worked fine
- IDE features for referencing projects failed with "Object reference" after first consumption of sigdata , and language features like intellisense,tooltips, colorization stopped working

There was no obvious clue about the real error coming from a different project.
Since regular build worked, the error creation went trough CI just fine and it only started to make troubles in IDE scenarios (= days later, different users of the project).

The following has been done:

- Prevent IDE error in the first place by deduplicating files and keeping the first occurence
- Report warning diagnostics in case of fsc.exe build when a duplicate is detected
- Report warning in case of service build
- Fix bug in componenttest framework for .fsi compilation which have been there and caused some tests to turn .fsi into .fs (therefore testing totally different files)

Evidence:
image

### CODE REVIEW

**Review by vzarytovskii — APPROVED** [review]

File: tests/FSharp.Test.Utilities/ProjectGeneration.fs [hunk 1]
```diff
@@ -378,7 +384,6 @@ type WorkflowContext =
 
 let SaveAndCheckProject project checker =
     async {
-        validateFileIdsAreUnique project
```
Comment (0101) [inline tests/FSharp.Test.Utilities/ProjectGeneration.fs:381]: This was there to ensure "fileId" is a unique identifier. Might lead to some issues if we relax that...

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/Structure/UnionTypeWithSignature02.fs [hunk 2]
```diff
@@ -2,7 +2,7 @@
 // Regression test for FSHARP1.0:4040
 // "Signature files do not prevent compiler-generated public constructors from leaking out of discriminated unions"
 // Note that the corresponsing .fsi file is NOT missing the "| C of int" part of the DU
-namespace N
+module N
```
Comment (0101) [inline tests/FSharp.Compiler.ComponentTests/EmittedIL/Structure/UnionTypeWithSignature02.fs:5]: Why this change?

File: tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs [hunk 3]
```diff
@@ -89,3 +89,13 @@ let ``Changes in a referenced project`` () =
...
     }
+
+[<Fact>]
+let ``Language service works if the same file is listed twice`` () = 
+    let file = sourceFile "First" []
+    let project =  SyntheticProject.Create(file)
+    project.Workflow {       
+        checkFile "First" expectOk
+        addFileAbove "First" file
+        checkFile "First" (expectSingleWarningAndNoErrors "Please verify that it is included only once in the project file.")
```
Comment (0101) [inline tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs:100]: Should we check the exact message to ensure the numbers are correct?

**Review by 0101 — APPROVED** [review]

File: tests/FSharp.Compiler.ComponentTests/EmittedIL/Structure/UnionTypeWithSignature02.fs (hunk 2 above)
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/EmittedIL/Structure/UnionTypeWithSignature02.fs:5]: Because the test was never executed (see last point of PR description) and wrong.

File: tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs (hunk 3 above)
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/FSharpChecker/CommonWorkflows.fs:100]: I see the numbers as a cosmetic detail which is only a small guide, not and essential part of this feature.

File: tests/FSharp.Test.Utilities/ProjectGeneration.fs (hunk 1 above)
Comment (T-Gro) [reply tests/FSharp.Test.Utilities/ProjectGeneration.fs:381]: Well this is in order to test the claim that IDE features do not start falling apart when this happens...
If real users can add the same file twice, the tests should be able to simulate it too :)

**Comment by T-Gro** [comment]
CompilerOptions\fsc\**gccerrors** (1) -- failed  ;; will check it.

**Review by 0101 — APPROVED** [review]

**Review by vzarytovskii — APPROVED** [review]
